zebra-state/src/service/check/nullifier.rs

```diff
@@ -0,0 +1,119 @@
+//! Checks for nullifier uniqueness.
+//!
+//! "A nullifier MUST NOT repeat either within a transaction,
+//! or across transactions in a valid blockchain.
+//! Sprout and Sapling and Orchard nullifiers are considered disjoint,
+//! even if they have the same bit pattern."
+//!
+//! https://zips.z.cash/protocol/protocol.pdf#nullifierset
+
+use std::collections::HashSet;
+
+use tracing::trace;
+
+use crate::{
+    error::DuplicateNullifierError, service::finalized_state::FinalizedState, PreparedBlock,
+    ValidateContextError,
+};
+
+/// Reject double-spends of nullifers:
+/// - one from this [`Block`], and the other already committed to the [`FinalizedState`].
```

Comment:
Minor: should this be

```suggestion
/// - one from this [`PreparedBlock`], and the other already committed to the [`FinalizedState`].
```